src/library/scala/concurrent/BatchingExecutor.scala

```diff
@@ -26,7 +26,13 @@ trait Batchable {
 
```

Comment:
@NthPortal @tarsa @SethTisue

This commit is a significant improvement in fairness and simplification of the code, with some dramatic performance improvements for synchronous/trampolining implementations since it allows for nested on-stack Runnable invocations (up to a limit of 16).

Reply:
@anatse @mkeskells @hepin1989 Please also have a look at this if you have time/interest! :)

Reply:
> with some dramatic performance improvements for synchronous/trampolining implementations since it allows for nested on-stack Runnable invocations (up to a limit of 16).

I recently tried out a similar optimization in the akka-streams `GraphInterpreter`. The main cost of the trampolining is the multimorphic dispatching to the thunks to execute next. I hoped that running it on the stack would enable the JIT to inline concrete implementations (with all the potentially positive ripple-on effects). Turned out that it didn't work out for the `GraphInterpreter` because inlining didn't happen. I tried to dig into it and came to the conclusion that the JIT would have to reevaluate some previous inlining decisions to recognize that some call-sites would now be monomorphic which it didn't do.

I think this may be a general result (but needs more investigation) that the JIT cannot optimize away the overhead of "sufficiently generic" code. The only way to remove those megamorphic call-sites (e.g. inside of any higher-order function that calls a user-supplied function like `map` or `flatMap` that cannot be inlined from the outside), is to provide more concrete single-use call-sites which is only possible by generating more code in the first place.

Reply:
>So, one advice for benchmarking generic things like `Future` is too make sure to have complex enough test cases. If you only test `future.map(f).flatMap(g)` then invoking those on the stack might lead to complete inlining in the best case. So, the test needs at least use three different arguments to `map` etc to make those megamorphic.

Well, that would not necessarily be anything else but a worst-case benchmark, wouldn't it? (Not all callsites will be megamorphic).

Reply:
@jrudolph 

>I think this may be a general result (but needs more investigation) that the JIT cannot optimize away the overhead of "sufficiently generic" code. The only way to remove those megamorphic call-sites (e.g. inside of any higher-order function that calls a user-supplied function like `map` or `flatMap` that cannot be inlined from the outside), is to provide more concrete single-use call-sites which is only possible by generating more code in the first place. I tried that out manually for the main akka-http Graph [here](https://github.com/jrudolph/akka/commit/df88ddf20dc1b305cfce6dc99714ebf12d578a63#diff-20ce84dc90488608e4a75c28b5ed0dbdR834) and found some great speed improvements (but how would you generate those classes automatically? Tricky stuff...).

Hehe, that code is pretty wild. :-)

Reply:
> Well, that would not necessarily be anything else but a worst-case benchmark, wouldn't it? (Not all callsites will be megamorphic).

Not sure if I'd call a code-base that uses Future.map more than once or twice worst-case... ;)

Reply:
@jrudolph But it's per *call-site* not per method invocation.

Reply:
Not sure what you mean.

What I meant is that if you use `Future.map(f)` with different arguments then the call-site to call f` will be megamorphic in the `map` implementation (i.e. in https://github.com/viktorklang/scala/blob/50c6f945f16926e1aab8ddd618a71aec68438156/src/library/scala/concurrent/impl/Promise.scala#L431). If you just measure Future infrastructure cost, I suspect the cost of this dispatch will be significant. I suspect more significant than switching from trampolining (BatchingExecutor) to running it on the stack with this change here. I can't say for sure without running the benchmarks myself, just saying that adding a benchmark that is more complex (several `Future.map` calls with different arguments) will show more realistic results.

Reply:
Add a bench to FutureBenchmark which illustrates what you mean and then let's see how it performs :)